### PR TITLE
Specialize the residual-style plots for histogram data

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Improve the display of residual-style plots for integrated datasets. Fix https://github.com/sherpa/sherpa/issues/1817.

# Details

This is a rework of #1989 without the changes in #1988 (which involve the class structure and not really the functionality of the code).

This introduces Data1DInt and DataPHA variants of the resid/delchi/ratio/chisqr plots. It follows the approach taken in

- https://github.com/sherpa/sherpa/pull/931

(and perhaps should have been included in that PR). The main changes are 

- adding the new classes for residual+histogram data
- adding in support for them to the UI layer

However, as part of this there are some other clean ups and tweaks, including

- 814764cc which cleans up the component hierarchy as it looks like that was either written before some classes were added or the original author never realized we had more-appropriate superclasses (this doesn't change functionality, but rather makes things a little-bit clearer about what being a "component" class is). in particular, we avoid having multiple superclasses which mixed up histogram/Data1DInt and point/Data1D behaviour as we didn't gain anything from this.
- bfe07232 just adds a minor refactoring of some `DataPHA` plot code to encapsulate the logic for creating the X axis values
  - this should re-appear below 
- 2ba5ca05 just adds a check to ensure we send a DataPHA object to the plot code that accepts it (which just makes the error message if you ever send in the wrong object nicer, and will be helpful if we add type rules to this module, which is not done here as that leads into #1988 which we don't need here)
- 01e3f992 takes the existing resid/ratio/delchi/chisqr plot code and creates a new class that the `ResidPlot`/... classes inherit from
  - the old code had some idea of separating out the common behaviour from the class-specific, but it was done in such a way that we had to repeat the common code for each class. This approach now makes the common code actually be common and moves the class-specific logic into the classes
  - this approach is about to be repeated
-  8c6502eb is the actual main part of this PR as it adds in histogram-specific residual plots
  - this follows the approach I just added for the Data1D class 
  - it's a bit-more complex than the Data1D case as we allow the X values to be overloaded (as well as the y / yerr / labels values) since we want to take advantage of the DataPHA specialization I added earlier in bfe07232
  - it adds in code for the UI layer (so that `plot_resid` will use the correct class)
  - it cleans up a little bit of the test code as the behaviour is a bit-more self consistent (but this is mainly only visible to me as the person who originally wrote the tests in question)
- 7ba47bc2 repeats the process for the existing contour residual classes, although it's easier as we only have resid and ratio to care about here (ie no delchi or chisqr versions)
- ba2a60f4 pylint pointed out to me that the new `MultiPlot` class inherited from `NoNewAttributesAfterInit` but never actually called it, so I decided to "fix" this by moving this class over to using slots instead (a la #2008)
- 06ffa195 is just an attempt to improve the test coverage of this PR (thanks to the way the code is written there are some lines that are just not worth testing - e.g. the methods in the new `BaseResidualXXX` classes that have to be over-ridden, and some lines that got changed here but don't have existing tests) but may as well try to catch some simple cases
- 024820c8 just makes sure that the histogram residual plots display Y errorbars by default

There's definitely some bike-shedding (e.g. the names of these "intermediary" residual classes I add) and should they be included in the RTD documentation.

# Example

## Data1DInt

With the script

```python
import matplotlib.pyplot as plt
from sherpa import ui

ui.load_arrays(1, [1, 2, 4, 8], [2, 3, 7, 9], [2, 4, 22, 9], ui.Data1DInt)
ui.set_source(ui.polynom1d.poly)
poly.c0 = 1
poly.c1 = 1

ui.plot_resid()
plt.savefig("resid.png")

ui.plot_ratio()
plt.savefig("ratio.png")

ui.plot_delchi()
plt.savefig("delchi.png")

ui.plot_chisqr()
plt.savefig("chisqr.png")
```

we get

### CIAO 4.16

resid

ratio

delchi

chisqr

### This PR

resid

ratio

delchi

chisqr

## DataPHA

With the script

```python
import matplotlib.pyplot as plt
from sherpa.astro import ui

ui.load_pha(1, "/home/dburke/sherpa/sherpa-1/sherpa-test-data/sherpatest/3c273.pi")
ui.ignore(hi=0.5)
ui.ignore(lo=7)

ui.set_source(ui.xsphabs.gal * ui.powlaw1d.pl)

ui.fit()

ui.plot_fit_resid(xlog=True, ylog=True)

plt.savefig("fit_resid.png")
```

which gets the fit (in both cases)

```
Dataset               = 1
Method                = levmar
Statistic             = chi2gehrels
Initial fit statistic = 5.43144e+10
Final fit statistic   = 20.7521 at function evaluation 259
Data points           = 40
Degrees of freedom    = 37
Probability [Q-value] = 0.985667
Reduced statistic     = 0.560868
Change in statistic   = 5.43144e+10
   gal.nH         0.00639106   +/- 0.135014    
   pl.gamma       1.9472       +/- 0.197551    
   pl.ampl        0.000184542  +/- 1.86246e-05 
```

### CIAO 4.16

fit_resid

### this PR

fit_resid

### CODE REVIEW

**Review by hamogu — CHANGES_REQUESTED** [review]
The plots shown in the description of the PR look good. And to some degree that's what matters for users - if the answer is right, it's right. I've left a few comments on the code structure. I think we make our lives unnecessarily complicated by too many levels of indirection, e.g. I think we can just call `backend.hline` when we need it instead of calling a `haling`method in the superclass, which then calls a functions, that calls `backend.hline`. That just makes the logic unnecessarily hard to follow; I suspect (but maybe I'm missing some intricacies because it's hard to follow) that the same is true for `title`, `_title`, and `_calc_title` etc. 
Let's just write what we do.

Further, we do need to look into the class structure. The more plot classes we add, the more repreated and boiler-plate code we add, that could hopefully be addresses with changes to the class structure. However, I fully agree to get this PR merged now and discuss that in a different place. In fact, having this PR merged first, will help us see how we need ot change the class structure, so let's just get this one in.

**Review by anetasie — APPROVED** [review]
Is the new residual plot going to be a default one now? Is there a preference setting for not having the histogram line drawn on the residual points? or remove the points and draw only a line?

File: sherpa/astro/ui/utils.py [hunk 1]
```diff
@@ -11380,6 +11385,74 @@ def get_source_component_plot(self, id, model=None, recalc=True):
...
+        if isinstance(data, DataPHA):
+            plotobj = self._plot_types["chisqr"][2]
+            if recalc:
+                plotobj.prepare(data, self.get_model(id), self.get_stat())
+
+            return plotobj
+
+        return super().get_chisqr_plot(id, recalc=recalc)
+
+    get_chisqr_plot.__doc__ = sherpa.ui.utils.Session.get_chisqr_plot.__doc__
```
[the remark this answers is not in the document]
Comment (hamogu) [reply sherpa/astro/ui/utils.py:11454]: We'll leave that for another PR.

File: sherpa/plot/__init__.py [hunk 2]
```diff
@@ -735,6 +739,24 @@ def plot(self, overplot=False, clearwindow=True, **kwargs):
...
+
+    @staticmethod
+    def hline(y, xmin=0, xmax=1,
+              linecolor=None, linestyle=None, linewidth=None,
+              overplot=False, clearwindow=True):
+        "Draw a line at constant y, extending over the plot."
+        backend.hline(y, xmin=xmin, xmax=xmax, linecolor=linecolor,
+                      linestyle=linestyle, linewidth=linewidth,
+                      overplot=overplot, clearwindow=clearwindow)
+
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/plot/__init__.py:759]: So, I am just trying to mirror the existing code here. I started on doing a full-scale refactoring in #1988 but, as you noted, there are a lot of questions to think about there and I just want to get some PRs in that actually help the user!

File: sherpa/plot/__init__.py [hunk 3]
```diff
@@ -1163,8 +1191,9 @@ def plot(self, overplot=False, clearwindow=True, **kwargs):
         # Note: the user arguments are not applied to the vertical line
         #
         if self.lr <= self.xhi.max() and self.lr >= self.xlo.min():
-            Plot.vline(self.lr, linecolor="orange", linestyle="solid",
-                       linewidth=1.5, overplot=True, clearwindow=False)
+            HistogramPlot.vline(self.lr, linecolor="orange",
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/plot/__init__.py:1194]: See comment above

File: sherpa/plot/__init__.py [hunk 4]
```diff
@@ -2178,7 +2238,227 @@ def contour(self, overcontour=False, clearwindow=True, **kwargs):
...
+        kwargs.pop('ylog', True)
+
+        # The superclass is not called since the y errors are
+        # potentially included in the plot.
+        #
+        Plot.plot(self, x, y, yerr=self.yerr, xerr=self.xerr,
+                  title=self.title, xlabel=self.xlabel,
+                  ylabel=self.ylabel, overplot=overplot,
+                  clearwindow=clearwindow, **kwargs)
+        super().hline(y=self.residual_axis, xmin=0, xmax=1,
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/plot/__init__.py:2316]: See comment above

**Comment by anetasie** [comment]
I reviewed the PR and approved it. I have one question about the preference of setting the histogram line. Is the preference setting allowing for the three options?
- plotting only data points
- data and histogram line overplotted
- only a histogram line and no data points.

File: sherpa/plot/__init__.py [hunk 5]
```diff
@@ -2206,31 +2486,54 @@ class DelchiPlot(ModelPlot):
...
 
-        self.y = self._calc_delchi(y, staterr)
-        self.yerr = staterr / staterr
+        self.y = (ylist[0] - ylist[1]) / staterr
+        self.yerr = np.ones_like(self.y)
+
+    def _change_ylabel(self) -> None:
         self.ylabel = 'Sigma'
+
+    def _title(self, data: Data1D) -> None:
```
[the remark this answers is not in the document]
Comment (DougBurke) [reply sherpa/plot/__init__.py:2507]: So, there's a lot of things that we could do to address this, but do do it generically would just slow down getting something useful to the user. I am getting a bit tired of looking at an area of Sherpa, deciding to rework it, and then getting bogged down in trying to work out how to pay off the technical debt. So I'm trying to focus on user-facing changes since we don't have much time to the next release!

**Comment by DougBurke** [comment]
For @anetasie - you can see the default behaviour for PHA data above in the `Examples / PHA` case.

This works just like the existing plots - so for the PHA case we now have residual plots that follow the data/model plots.

So we have, for a PHA dataset with the matplotlib backend:

```python
plot_resid()
```

sherpa1

```python
plot_resid(linestyle="none")
```

sherpa2

```python
plot_resid(linestyle="none", marker=".")
```

sherpa3

```python
plot_resid(marker=".", linewidth=1)
```

sherpa4

```python
plot_resid(linewidth=1, yerrorbars=False)
```

sherpa5


Now, it gets complicated when you get to `plot_fit_resid` since there's no way to say "only apply this to the residual plot", so you get

```python
plot_fit_resid(xlog=True, ylog=True)
```

sherpa6

In the following the model line disappears, which is not ideal....

```python
plot_fit_resid(xlog=True, ylog=True, linewidth=1, linestyle="none")
```

sherpa7

You can try

```python
plot_fit_resid(xlog=True, ylog=True, linewidth=1, linestyle="none", marker=".")
```

sherpa8

but at this point the simple plot interface breaks down and you'd have to do things a bit differently.

File: sherpa/plot/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hamogu) [reply sherpa/plot/__init__.py:2507]: That's fine. We're not making it worse; you're just following the patterns that that were established long ago. In fact, I think it's no me to re-work this, I've just not done that yet. So, agreed, let's get this to the user and worry about technical dept some other time.

**Review by hamogu — APPROVED** [review]
I've raised a few point I noticed in code review, but after discussion I agree that paying off old technical dept is not the goal of this PR, so I'm approving now.

**Comment by anetasie** [comment]
@DougBurke Thanks for the example plots! This looks good.
